### PR TITLE
Reimplement strtod and parseFloat / strtod + Fix String.fromCodePoint + improve strtol

### PR DESCRIPTION
**Pull request by MaxGraey**
- [x] Improve precision and parsing for `parseFloat` / `strtod`;
- [x] Fix `String.fromCharCode`;
- [x] Minor improvements for `parseInt` / `strtol`.

**DONE**

### CODE REVIEW

**Comment by MaxGraey** [comment]
@dcodeIO I'm not very happy with this implementation. It has some rouge issue with precision (see commented tests). But it definitely better than previous one. Just keep in mind it will be improved in future

**Comment by MaxGraey** [comment]
This also done. Have some rouge edges (which will be improved later) but definitely better solution than previous one.

File: std/assembly/index.d.ts [hunk 1]
```diff
@@ -1467,6 +1467,8 @@ interface INativeMath<T> extends IMath<T> {
   sincos_cos: T;
   /** Seeds the random number generator. */
   seedRandom(value: i64): void;
+  /** Multiplies a floating point `x` by 2 raised to power exp `n`. */
+  scalbn(x: T, n: i32): T;
```
Comment (dcodeIO) [inline std/assembly/index.d.ts]: Is this useful to a user as a public API?

File: std/assembly/string.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ import { idof } from "./builtins";
   @lazy static readonly MAX_LENGTH: i32 = BLOCK_MAXSIZE >>> alignof<u16>();
 
   static fromCharCode(unit: i32, surr: i32 = -1): string {
+    assert(unit >= 0);
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Seems JS simply converts this to unsigned:

```ts
> String.fromCharCode(-123123)
'Ἅ'
> String.fromCharCode(-123123 >>> 0)
'Ἅ'
```

File: std/assembly/index.d.ts (hunk 1 above)
Comment (MaxGraey) [reply std/assembly/index.d.ts]: I guess yes. It's has different name in C/C++ as `frexp`

File: std/assembly/string.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: Yeah. Will improve this

File: std/assembly/util/string.ts [hunk 3]
```diff
@@ -160,54 +180,236 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+
+// @ts-ignore: decorator
+@inline
+function scientific(significand: u64, exp: i32): f64 {
+  if (!significand || exp < -342) return 0;
+  if (exp > 308) return Infinity;
+  // Try use fast path
+  var result = strtodFast(<f64>significand, exp);
+  if (!isNaN(result)) {
+    trace("fast path:", 1, result);
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: left over

File: std/assembly/util/string.ts [hunk 4]
```diff
@@ -160,54 +180,236 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+    return result;
+  }
+  if (exp < 0) {
+    result = scaledown(significand, exp);
+    trace("scaledown:", 1, result);
+    return result;
+  } else {
+    result = scaleup(significand, exp);
+    trace("scaleup:  ", 1, result);
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: left over

File: std/assembly/util/string.ts [hunk 5]
```diff
@@ -160,54 +180,236 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+  if (exp > 308) return Infinity;
+  // Try use fast path
+  var result = strtodFast(<f64>significand, exp);
+  if (!isNaN(result)) {
+    trace("fast path:", 1, result);
+    return result;
+  }
+  if (exp < 0) {
+    result = scaledown(significand, exp);
+    trace("scaledown:", 1, result);
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: left over

File: std/assembly/util/string.ts [hunk 6]
```diff
@@ -160,54 +180,229 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+  return copysign<f64>(scientific(x, position - min(capacity, consumed) + parseExp(ptr, len)), sign);
+}
+
+// @ts-ignore: decorator
+@inline
+function scientific(significand: u64, exp: i32): f64 {
+  if (!significand || exp < -342) return 0;
+  if (exp > 308) return Infinity;
+  // Try use fast path
+  var result = strtodFast(<f64>significand, exp);
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: Wondering if it would make sense to move the logic of `strtodFast` to here, saving the check for `isNaN`, doing the scaling in the `else` case.

File: std/assembly/util/string.ts [hunk 7]
```diff
@@ -160,54 +180,229 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+  for (; exp >= 13; exp -= 13) {
+    significand = fixmul(significand, coeff);
+  }
+  significand = fixmul(significand, <u32>ipow32(5, exp));
+  shift = __fixmulShift;
+  return NativeMath.scalbn(<f64>significand, <i32>shift);
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h#L27
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: Seems this reference should be `src/stdlib/parse/float.h#L8`. Also wondering if the line numbers can lead to more confusion than being helpful when metallic source code updates.

**Comment by dcodeIO** [comment]
Looking great overall, just a few remaining comments :)

File: std/assembly/util/string.ts (hunk 6 above)
Comment (MaxGraey) [reply std/assembly/util/string.ts]: Good catch!

File: std/assembly/util/string.ts (hunk 7 above)
Comment (MaxGraey) [reply std/assembly/util/string.ts]: Yeah, line details are unnecessary

**Comment by MaxGraey** [comment]
@dcodeIO I also added some tests from tc39/test262 and found one issue. Plz don't merge this

**Comment by MaxGraey** [comment]
Ok. Now should be more right. I switched to approach for handling `Pow10` as in RapidJSON with division on `significand / pow10(-exp)` when `exp` is negative instead multiply it to negative power from LUT. This approach don't loose precision in some rare cases.

cc @dcodeIO
plz review it again)

**Comment by MaxGraey** [comment]
It ready to re-review

File: std/assembly/util/string.ts [hunk 8]
```diff
@@ -160,54 +167,211 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
...
+
+// @ts-ignore: decorator
+@lazy
+var __fixmulShift: u64 = 0;
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h
+// @ts-ignore: decorator
+@inline
+function fixmul(a: u64, b: u32): u64 {
```
Comment (dcodeIO) [inline std/assembly/util/string.ts]: This is called twice in scaleup. No strong opinion, but since inlined this would exist twice even if optimizing for size.

Comment (MaxGraey) [reply std/assembly/util/string.ts]: Yes, I prefer to leave this inlined because it call in loop and in my experience this potentially could give significant speed boost especially because this touches global variable internally

File: tests/compiler/std/string.ts [hunk 9]
```diff
@@ -80,27 +81,206 @@ assert("ab c".trim() == "ab c");
...
+assert(isNaN(parseFloat("+Infinit")));
+assert(isNaN(parseFloat("infinity")));
+
+// precision tests
+// 2^-1074 * 0.5 - eps
+assert(parseFloat(".2470328229206232720882843964341106861825299013071623822127928412503377536351043e-323") == 0);
+
+// 2^-1074 * 0.5 + eps
+// const Ox1p_1074 = reinterpret<f64>(0x1); // 0x1p-1074
+// assert(parseFloat(".2470328229206232720882843964341106861825299013071623822127928412503377536351044e-323") == Ox1p_1074); // FAIL
```
Comment (dcodeIO) [inline tests/compiler/std/string.ts]: What's the implication of these cases failing?
* 2^-1074 * 0.5 + eps
* 2^-1074 * 1.5 - eps
* 2^1024 - 2^970

Comment (MaxGraey) [reply tests/compiler/std/string.ts]: Will add comment

Comment (MaxGraey) [reply tests/compiler/std/string.ts]: For 2^1024 - 2^970 we have already comment btw

**Comment by dcodeIO** [comment]
Regarding these failure cases: How bad is this practically? Are these merely edge cases?

**Comment by MaxGraey** [comment]
It has only corner cases for extremely large or small values. To improve this cases we should choose much more complex solution which I'll definitely implement later. But this solution is pretty enought for most cases. See [this comment](https://github.com/jdh8/metallic/issues/5#issuecomment-524296508). And this affect only for when `abs(x) >= ~1e40`. We could potentially have wrong 2 last digits

**Comment by dcodeIO** [comment]
I see, thanks. So this is due to limiting to 64 bits for intermediate values, right? If so, tempted to say that this might be as good as it gets at this point in time anyway, since emulating 128 bit floats without multi-value looks like it'd need a shadow stack / heap support anyway. Even if metallic did that with structs, we'd probably still want to use multi-value.

**Comment by MaxGraey** [comment]
128-bit soft floats usually slower than 128-bit fixed points. So 128-bit modular algo much better and use for such cases pretty widely for double <-> string conversion and trig argument reduction (Payne Hanek approach for large arguments)

**Comment by MaxGraey** [comment]
@dcodeIO Currently it's not absolutely ideal. I'll rewrite part for handling huge values but this require some refactoring because `dtoa` also based on same BigFloat routines so will be great unify all of this and the same time fix some mispricisions on `dtoa`. But could be done later I guess.

**Comment by dcodeIO** [comment]
Thanks!
